### DISCUSSION TITLE
希望大佬添加一下对alist播放视频的自动识别

### DISCUSSION THREAD

**Discussion by truewhile** [Mr-Quin/danmaku-anywhere#38] [General]
alist播放视频格式比较固定，应该可以拿到播放视频的名称，希望大佬添加一下对alist视频播放的自动识别

**Comment by Mr-Quin** [comment]
自动适配这块打算重做成基于自定义策略的，用户可用选择器编写策略来适配自己的网站

**Comment by truewhile** [comment]
我看了一下这个class名称应该是固定的吧，让用户填写class名称去匹配播放文件名称是不是会比较好
![image](https://github.com/user-attachments/assets/981e6879-d267-4ae0-81d9-93b7b89999d7)
![image](https://github.com/user-attachments/assets/56cb8b67-4989-41aa-8cf8-07822401564a)

**Comment by truewhile** [comment]
我不是前端，我不知道，前端打包后某一代码块的class名称是不是一直保持不变，如果class是保持不变并且唯一的，那class名称去匹配是不是比较好

**Comment by Mr-Quin** [comment]
`class`不一定是固定的，只用`class`的话不够普适
现在尝试的是XPath，可以覆盖`class`,`id`,`tag`

以AList举例
![image](https://github.com/user-attachments/assets/6ed85ea1-5396-4b0d-90a7-5c77bdbc8aa5)
从这里可以提取XPath `//*[@id="hope-select-cl-40-trigger"]/div`获得文件名

然后再用regex从文件名提取标题，集数等信息。这个可能性太多了，得根据自己的文件命名规则来填
![image](https://github.com/user-attachments/assets/f6d3bce9-73f3-436e-b991-89812a7885c4)

**Comment by Mr-Quin** [comment]
v0.15.0中添加自定义适配，可以尝试以下

**Comment by truewhile** [comment]
> v0.15.0中添加自定义适配，可以尝试以下
0.15.0好像构建失败了，这个标题识别是根据正则表达式吗？昨天晚上刚跟弹弹paly作者请教弹弹play是怎么识别的，到最后得出结论因为各翻译组命名格式都不一样，基本没办法用一个正则表达式去识别，只能针对某一个翻译组去做识别，或者是使用ai去做识别也行，弹弹play作者建议我直接调用他的开放接口根据文件名称去获取番剧信息，我感觉这块识别是不是直接调用弹弹play作者的接口比较好，用户会不会写正则表达式是一说，关键正则的匹配率也不高，不如直接调用弹弹play的开放接口识别率高，https://api.dandanplay.net/swagger/ui/index
，这个是弹弹play作者给我的接口文档里面的/api/v2/match接口感觉很适合这个场景，如果能获取播放视频元素的前16MB MD5数据识别率会高很多

**Comment by truewhile** [comment]
````html
<video id="myVideo" controls>
  <source src="your-video-file.mp4" type="video/mp4">
</video>

<script src="https://cdn.jsdelivr.net/npm/crypto-js@4.1.1/crypto-js.js"></script>

<script>
  async function getVideoMD5() {
    const video = document.getElementById('myVideo');
    const videoSrc = video.querySelector('source').src;

    // 获取视频文件的前16MB数据
    const response = await fetch(videoSrc);
    const reader = response.body.getReader();
    
    let receivedLength = 0; // 当前接收到的字节数
    let chunks = []; // 保存流的字节块
    let chunkSize = 16 * 1024 * 1024; // 16MB
    
    while (receivedLength < chunkSize) {
      const { done, value } = await reader.read();
      if (done) break;
      chunks.push(value);
      receivedLength += value.length;
    }

    // 组合所有片段
    const blob = new Blob(chunks);
    
    // 读取 Blob 内容为 ArrayBuffer
    const arrayBuffer = await blob.arrayBuffer();
    
    // 使用 CryptoJS 生成 MD5 哈希
    const wordArray = CryptoJS.lib.WordArray.create(arrayBuffer);
    const md5Hash = CryptoJS.MD5(wordArray).toString();
    
    console.log('MD5 Hash:', md5Hash);
  }

  // 监听页面加载完后调用
  document.addEventListener('DOMContentLoaded', getVideoMD5);
</script>
````
这个是chatgpt生成的获取video元素前16MB内容生成MD5的代码，大佬可以参考一下，我感觉还是使用获取哈希+标题去请求弹弹play的接口为主，手写正则表达式为辅比较好，用户手写正则表达式难度太大了

**Comment by Mr-Quin** [comment]
构建显示失败是因为添加了新的权限所以商店要再审核

让用户写regex是有点难，但是我现在也想不到有什么更通用的办法。用一堆regex来提取标题什么的也不是没有，像这样https://github.com/Sonarr/Sonarr/blob/e6bd58453a532c137879c2b1a6a267dc4bf03828/src/NzbDrone.Core/Parser/Parser.cs#L20

弹弹Play有md5接口我是刚知道，这个可以研究下。但是毕竟这个插件不止支持弹弹play，其他源还是要靠关键词

我不清楚用这个插件的都是什么使用场景，我自己是Plex，Plex会对视频再次压制，而且视频标题也不会沿用文件名，而是根据tvdb等数据库重命名的。这种场景，还有比如Youtube之类的没有可用的文件名和哈希值的情况用弹弹Play的匹配也不一定效果很好。

Regex的匹配成功率的话，我自己这几天测试使用regex还没有出过问题，但是这个比较依赖标题的准确性。

现在的这个方案是最大化通用性，之后再对各个源单独做优化，比如md5。

单开了一个issue #35 追踪一下。可能可以做个开关让用户选择，我的理解能够使用md5的场景主要还是能够直接播放视频文件的情况，个人网盘之类的

**Comment by truewhile** [comment]
> 构建显示失败是因为添加了新的权限所以商店要再审核
> 
> 弹弹Play有md5接口我是刚知道，这个可以研究下。但是毕竟这个插件不止支持弹弹play，其他源还是要靠关键词
> 
> 我不清楚用这个插件的都是什么使用场景，我自己是Plex，Plex会对视频再次压制所以md5应该是匹配不到的，而且视频标题什么也不会沿用文件名，而是的是根据tvdb等数据库重命名的，有时候和bangumi什么的也对应不上。
> 
> 现在的这个方案是最大化通用性，之后再对各个源单独做优化，比如md5。
> 
> 让用户写regex是有点难，但是我现在也想不到有什么更通用的办法。用一堆regex来提取标题什么的也不是没有，像这样 https://github.com/Sonarr/Sonarr/blob/e6bd58453a532c137879c2b1a6a267dc4bf03828/src/NzbDrone.Core/Parser/Parser.cs#L20

这正则表达式也太地狱了
![image](https://github.com/user-attachments/assets/acafe2af-34b3-4ffe-bfb6-585385b589bb)
我也是自己在写个小工具才去研究这个，我在写订阅ssr推送到115网盘离线下载的时候，想做个根据下载文件名称自动分类的功能，才研究这个东西，网上有人写了根据标题获取番剧名称的python库，但是识别率不是很高，昨天才去请教弹弹play的作者的，然后我那个是推送到115进行离线下载，md5的方案行不通，只能根据文件名去识别一下
![image](https://github.com/user-attachments/assets/428871b9-255a-4247-9cb7-13f2798c5d5b)
正则表达式让chatgpt帮我写了写，发现效果最差就没深入研究了
![image](https://github.com/user-attachments/assets/505ea6a9-af20-4c75-8cba-b89280c41ea1)

**Comment by Mr-Quin** [comment]
这个Sonarr我也在用，他就是用这些regex来识别番剧名然后再和数据库匹配，识别率还是不错的，很少碰到问题。其实如果视频来源比较固定，命名比较规则，regex应该还是可以的。我因为工作关系经常写，所以还好。网上有很多regex生成器也可以试试。

**Comment by truewhile** [comment]
plex,emby这些都没办法挂载网盘资源或者webdav资源，不然我也用了，我现在已经不想下载到本地看了，我弄监听rss推送115就是准备后面不下载到本地看了，组nas的用 plex,emby的比较多，我感觉还是网盘比较好用，一些旧的种子什么的本地根本下不动，除非玩pt有人保种，不用emby或者plex这种工具就会导致文件名全是字幕组定义好的，一家一个命名规则，难顶

**Comment by Mr-Quin** [comment]
我看了一下，弹弹play的match api好像只提供文件名也可以匹配
从网页上计算哈希值和获取文件大小还是太复杂了，先从文件名开始吧
 #35 追踪

**Comment by truewhile** [comment]
> 我看了一下，弹弹play的match api好像只提供文件名也可以匹配 从网页上计算哈希值和获取文件大小还是太复杂了，先从文件名开始吧 #35 追踪

对，但是哈希值是必填的，可以去md5网站随便生成一个哈希值就行，用文件名称匹配准确率还行
![image](https://github.com/user-attachments/assets/3a1c62a1-9949-42fa-bc79-8e8fb4af1198)

**Comment by truewhile** [comment]
![image](https://github.com/user-attachments/assets/e74cf4d7-d166-4d26-aa72-7f2fca085388)
没有保存按钮，无法固定住，出去查个东西写的东西就没了然后所有东西都要充头开始配

**Comment by Mr-Quin** [comment]
啊这
![PoFacepalmGIF](https://github.com/user-attachments/assets/ece93cc2-a499-4b58-a379-bd27fc36723c)


保存在右上角
![image](https://github.com/user-attachments/assets/c0c2d89a-379e-4fce-9edf-2beb1206528c)
我一直用的深色模式没发现这个问题

**Comment by Mr-Quin** [comment]
开了 #36

**Comment by truewhile** [comment]
笑死我了，不过正则表达式真的写不来，我把你写的那个事例给chatgpt让它帮我写，它写出来的基本就没法用，[SZW] Strike the Blood TV13 [BDRpi][1080P][x.264 AAC][CHS-IN].mp4，这个我折腾半天，都没弄好，有个弄好了识别的是英文，也没出来结果

**Reply by Mr-Quin** [reply]
这个可以用
^\[(?<subgroup>[^\]]+)\](?<title>[^\]]+?) TV(?<episode>\d+) \[.*

我把Sonarr的一串regex移植过来吧

**Reply by truewhile** [reply]
> 这个可以用 ^[(?[^\]]+)](?<title>[^\]]+?) TV(?\d+) [.*
> 
> 我把Sonarr的一串regex移植过来吧

是不是能在官网加个页面专门存放各字幕组的正则表达式规则，毕竟用户里面有菜鸡也有大佬，大佬手搓完正则表达式提交到官网还能造福菜鸡，可以让用户提交后你那边审批一下，通过的话就放上面，当成一个库来用，用户在选规则的时候可以选择官方正则库或者自己手搓，这样易用性也提高了，只要后面用户越多那这个软件的易用性就越强

**Reply by Mr-Quin** [reply]
我的用户量可能还没到这个程度😂
不过有打算在编辑器里加一个“预设列表”或者模板之类的，大佬可以开pr到预设里
